config/snapshot/config_snapshot_test.go

```diff
@@ -108,6 +108,96 @@ func TestObserve_OnInit(t *testing.T) {
 	}
 }
 
+func TestObserve_OnInitRetry(t *testing.T) {
+	tests := []struct {
+		name             string
+		config           *configpb.Configuration
+		configGeneration int64
+		openErr          error
+		openOnRetry      bool
+		expectInitial    *configpb.Dashboard
+		expectError      bool
+	}{
+		{
+			name: "Reads config on retry",
+			config: &configpb.Configuration{
+				Dashboards: []*configpb.Dashboard{
+					{
+						Name: "dashboard",
+					},
+				},
+			},
+			configGeneration: 1,
+			expectInitial: &configpb.Dashboard{
+				Name: "dashboard",
+			},
+			openErr:     errors.New("fake error"),
+			openOnRetry: true,
+		},
+		{
+			name: "Returns error if config isn't present on retry",
+			config: &configpb.Configuration{
+				Dashboards: []*configpb.Dashboard{
+					{
+						Name: "dashboard",
+					},
+				},
+			},
+			configGeneration: 1,
+			expectInitial: &configpb.Dashboard{
+				Name: "dashboard",
+			},
+			openErr:     errors.New("fake error"),
+			expectError: true,
+		},
+	}
+
+	path, err := gcs.NewPath("gs://config/example")
+	if err != nil {
+		t.Fatal("could not path")
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			client := fakeClient()
+			client.Opener.Paths[*path] = fake.Object{
+				Data: string(mustMarshalConfig(test.config)),
+				Attrs: &storage.ReaderObjectAttrs{
+					Generation: test.configGeneration,
+				},
+				OpenErr:     test.openErr,
+				OpenOnRetry: test.openOnRetry,
+			}
+			client.Stater[*path] = fake.Stat{
+				Attrs: storage.ObjectAttrs{
+					Generation: test.configGeneration,
+				},
+			}
+
+			snaps, err := Observe(ctx, nil, client, *path, nil)
+
+			if err != nil {
```

Comment:
nit: You want to fail if these don't match, right? So you'll want to either check both cases (including "we expected an error, but didn't get one), or a more complex check like `if (err != nil) != test.expectError`

Reply:
Ah yup, fixed it up to check both cases. (iirc we do have unit tests in other components that do this if/else style when expecting errors.)